### ISSUE TITLE
git commands in the vscode web do not work, permission denied

### ISSUE THREAD

**Issue by vanarok** [loft-sh/devpod#711]
**What happened?**  
when devpod up git@git.sr.ht:~vanarok/asia-road-logistics-tracking and use git, i get: 
```
> git pull --tags origin master
git@git.sr.ht: Permission denied (publickey,keyboard-interactive).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

**What did you expect to happen instead?**  

Git commands working.

**How can we reproduce the bug?** (as minimally and precisely as possible)  

1. Create workspace with private repository
2. Run ```git pull``` from command palette in vscode web

**Local Environment:**  
- DevPod Version: master branch 19 september
- Operating System: Android OS with termux
- ARCH of the OS: ARM64

**DevPod Provider:**  
- Cloud Provider: digitalOcean
- Local/remote provider: ssh 
- Custom provider: [devpod-provider.txt](https://github.com/loft-sh/devpod/files/12726276/devpod-provider.txt)

[devpod-debug-output.txt](https://github.com/loft-sh/devpod/files/12740357/devpod-debug-output.txt)

**Comment by 89luca89** [comment]
hi @vanarok 

Can you try doing a `devpod ssh` to the workspace, and then (from vscode web) try again?

**Comment by vanarok** [comment]
I ran into other problems using devpod, so I'll come back to this issue a little later

**Comment by pascalbreuninger** [comment]
@vanarok did you have the chance to look at this again?

**Comment by rodoviario** [comment]
> hi @vanarok
> 
> Can you try doing a `devpod ssh` to the workspace, and then (from vscode web) try again?

I have similar issues, and I did tryed to devpod ssh into the running container. I can confirm that:
* In the external terminal, after the devpod ssh to the container, the git commands I issue work
* In the vscode-web terminal the commands not work, even after a succesfull push from outside
* If I do ps -a inside the vscode-web terminal, there are 3 devpod occurrences on the list
* If I do that from the outside terminal after the devpod ssh command there are those 3 occurrences plus 2 more (5 in total)

Another thing I've noted from the logs is that during the startup of the workspace, it looks like devpod is copying the ssh keys that exist on the host to the enviroment... but when I am inside, git complains for credentials, there is no .ssh directory until git fails (and creates one)... and there are no copies of the keys.

I've tryed copying a key inside the vscode-web terminal, and is not detected... also tryed ssh-add the key, and it says there is no ssh-agent... started one with eval... to no avail.

Even tryed to modyfy the vendor, adding fields for the priv and public key on the provider.yml, along with commands to create the .ssh, create the key files, start the ssh-agent and add them with ssh-add, but without success so far

I don't know much about golang, but I'm very interested in help, testing possible solutions or workarounds, so please don't hesitate in asking to try things here... I am doing that every day anyway :) looking for a way to develop inside devpod, but with access to my remote git repos.

Thanks in advance, and thanks for this great project, I love what it already does

**Comment by pascalbreuninger** [comment]
I've just tested this with a simple workspace and the local docker provider and it worked alright for me.
Which version and os are you on @rodoviario? And is it an older workspace or a new one?

For debugging purposes, the output of `git config --global --list` in your vscode terminal should resemble something like this
```
user.name=Your Name
user.email=yourname@provider.tld
credential.helper=/usr/local/bin/devpod agent git-credentials --port 17613 
```

**Comment by rodoviario** [comment]
![git debug](https://github.com/loft-sh/devpod/assets/4974682/45a1995b-0bad-45b0-b6e2-d9f3142de34a)
Indeed, it states my name and email, and the devpod agent for git credentials port

My host is Linux Mint 21.2 x86_64 - Mate flavor
Kernel: 6.2.0-37-generic

The workspace is a private fork of https://github.com/microsoft/vscode-remote-try-php
Which I git cloned using:
`git clone git@github.com:xxxxxxxxxxxx/xxxxxxxxx.git`

After that I used the standard docker provider, selected my local git repo as a source for the workspace, and the vscode web as an ide. Everything works until I start to modify things inside vscode web and try to push changes. Even after seeing in de devpod gui the messages that my ssh credentials had been inyected, vscode will try (and fail) only creating the .ssh directory with a known_hosts file inside (but not my ssh certs).
If I try to do git push, it also fails inside the vscode terminal
![git-push-inside-vscode-web-terminal](https://github.com/loft-sh/devpod/assets/4974682/9fba2dc1-4a1c-44e0-a51f-81b989d938f3)

but if I do from outside:
`devpod ssh aoc-php`

![from-outside](https://github.com/loft-sh/devpod/assets/4974682/783a7895-c265-485e-88b8-695c8c5f352b)

I am attaching the output of `ps -A` from inside the web terminal

I can recreate the repo with the debug option in devpod if it helps
I can also make my repo public (it should be almost equal to the original)
![ps-a-inside-vscode-web-term](https://github.com/loft-sh/devpod/assets/4974682/de983173-7846-4904-aa4b-5d31dfd85920)

Hope any of this helps to sort out what is going on
Thanks in advance!

**Comment by pascalbreuninger** [comment]
hey @rodoviario, sorry for the late response. The reason you see this is because you're cloning the repo using SSH. The git credentials helper is only for HTTP authenticated repos, that's why it doesn't work in vscode-browser. Should be fixed by cloning via HTTP.
We are currently thinking about marking SSH cloning as experimental...

**Comment by rodoviario** [comment]
That clarifies a lot.

* I installed git credential manager
* Created a PAT
* Cloned my private fork of te repo with https with my user and the PAT
* Entered to devpod and created a workspace using local cloned repo as a base
* When I was inside vscode-web, I changed a file in the repo and pushed
* It opened a web window with a code and a link to github... 
   - I was expecting kind of no interaction needed for the push to work
* When I came back to the original browser window it told me that I didn't have permisions to push to the 'original' remote and if I wanted to create a fork...
* But the fact was that I started from a private fork of the repo, so I stopped the process there.

I guess I have to learn more about http(s) github authentication, PATs (needed because 2FA) and such.
If I make it work I will report here the success.

Thanks for the lead

**Comment by pascalbreuninger** [comment]
@rodoviario thanks for the detailed info. 

Are you starting the workspace on a local copy of the repo? 
Once you're authenticated there is no need to pre-clone it to your local machine, you should be able to enter the repo url in DevPod and it'll reuse your git credentials to clone

**Comment by rodoviario** [comment]
Just tryied that. If I use the https url from github as it is, right from the 'clone' button, devpod fails with this (debug is disabled because it is long output, and I am unable to save / nor copy it and paste here)
![without-username](https://github.com/loft-sh/devpod/assets/4974682/80f5e749-273a-4c48-93d4-7639cdf06721)

On the other hand, if I append my username plus the '@' simbol before github.com as suggested elsewere (and the way it worked localy) devpod stops complaining about the username, but fails with the password. No way for me to append the PAT at the creating workspace/clonning step
![with-username](https://github.com/loft-sh/devpod/assets/4974682/80b30a54-f5a9-4160-a98c-31fcfc7ac293)

**Comment by pascalbreuninger** [comment]
is your git credential manager configured properly?
You can test this for github by running 
`echo "protocol=https\nhost=github.com" | git credential fill`, it should print something along the lines of 
```
protocol=https
host=github.com
username=pascalbreuninger
password=YOUR_PAT
```

**Comment by rodoviario** [comment]
Tryed... 
`fatal: rehusando trabajar con campo host faltante en la credencial`
Which means `fata: refused to work with missing host field in credential`

**Comment by pascalbreuninger** [comment]
tbh, looks like more of a git credentials setup problem on your machine rather than DevPod. you use other means to pipe stdin to `git credential fill` but ultimately you'll need to have your credentials stored if you don't want to add `username@PAT` to every repo

**Comment by rodoviario** [comment]
Thanks for point me to the solution... it worked.
I used the new (github I guess) git credential-manager, that support 2FA (which after installed should be invoked with `git credential-manager store`
It works similar to `git credential manage fill` so I piped a file containing the 4 needed fields. I tryed this one first as you suggested, in my other machine, but even if it returned me the 4 needed fields correctly... didn't worked when cloning my private repo.
So with this 'new' I was able to store the credentials. Also I was logged in to github from previous attempts (this one have a special command to make a login) but to be sure I logged out the account
And it worked, as expected, with the github http cloning link and nothing more to be foretold or added interactivelly. It just cloned my private github repo, and from inside I was able to push right away.

Thanks for all the tips and clues
